[RFC][BE] assume error checking is on by default

c-p-i-o opened the pull request:
Summary:
Remove conditional MACRO `ENABLE_NCCL_ERROR_CHECKING` and assume that error checking is always on.
These checks were wrapped in a macro because older NCCL libraries didn't have the pre-requisite functions to do error checks. This check was put in several years ago.

Pull request #142023 adds a static_assert that NCCL version should be 2.7 or above for PyTorch to work.
NCCL 2.7 released about 2 years ago so it's relatively safe to assume that everyone has upgraded.

Assume that the world has all upgraded to later version of NCCL library.

Release note for PyTorch must specify that going forward, PyTorch will only work with NCCL version 2.7 and above.

Test Plan: Unit tests.

Differential Revision: D66672062




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

wconstab commented:
I think this PR is essentially masquerading a bigger assumption (NCCL >=2.4) under a smaller title (error checking).  Would be good to make a few things more explicit
- did we pick the right cutoff version?  (is 2.4 old enough?)
- should we add an explicit assertion so users get a clear signal they are using unsupported nccl rather than a linker error for missing error checking?

I'd vote for making a specific PR for the nccl version assertion, and get more eyes on that (also state why we feel 2.4 or 2.x is the right oldest cutoff version to pick). If we can get that landed, then this PR is a no-brainer on top.

c-p-i-o commented:
> I think this PR is essentially masquerading a bigger assumption (NCCL >=2.4) under a smaller title (error checking). Would be good to make a few things more explicit
> 
> * did we pick the right cutoff version?  (is 2.4 old enough?)
> * should we add an explicit assertion so users get a clear signal they are using unsupported nccl rather than a linker error for missing error checking?
> 
> I'd vote for making a specific PR for the nccl version assertion, and get more eyes on that (also state why we feel 2.4 or 2.x is the right oldest cutoff version to pick). If we can get that landed, then this PR is a no-brainer on top.

I made a pull request #142023 with a static_assert that NCCL version is 2.4 or above. 
I added the rationale for my thought:
1. 2.4 released 2 years ago.
2. Enabling error checking (this PR) should be beneficial to all and they are probably getting it anyway. 
For the stragglers who haven't upgraded, this might provide impetus to update/upgrade?

wconstab reviewed (approved):
LGTM. Please update this PR desc to reference the 2.4 PR.

c-p-i-o commented:
@pytorchbot rebase

oraluben commented:
Hi, I just found this when I was trying to fix some build failure against old nccl: #145719.

I'd like to give some feedback about the overall NCCL integrate policy.

We're currently relying on nccl 2.12 and it's not likely to be upgraded in the short term. While building pytorch, it's common that we found it only work with the bundled version (fixed once in #130419).

I'll be happy to see if minimal NCCL version is specified (in doc or implicit in code, like `static_assert`), so corresponding CI could be added, upgrade policy could be discussed, etc.

cc @wconstab @kwen2501

c-p-i-o commented:
> Hi, I just found this when I was trying to fix some build failure against old nccl: #145719.
> 
> I'd like to give some feedback about the overall NCCL integrate policy.
> 
> We're currently relying on nccl 2.12 and it's not likely to be upgraded in the short term. While building pytorch, it's common that we found it only work with the bundled version and older versions are usually broken (fixed once in #130419).
> 
> I'll be happy to see if minimal NCCL version is specified (in doc or implicit in code, like `static_assert`), so corresponding CI could be added, upgrade policy could be discussed, etc.
> 
> cc @wconstab @kwen2501

Yes, we added a static assert that NCCL version must be 2.7 or above in an earlier PR.
That PR landed successfully.
Now, in this PR, I was attempting to remove the `Ifdefs` that implied an older NCCL version.
Unfortunately, there's something nefarious going on that's causing some tests to fail.

c-p-i-o commented:
@pytorchbot rebase

c-p-i-o committed:
[RFC][BE] assume error checking is on by default (#141914)

Summary:
Remove conditional MACRO `ENABLE_NCCL_ERROR_CHECKING` and assume that error checking is always on.
These checks were wrapped in a macro because older NCCL libraries didn't have the pre-requisite functions to do error checks. This check was put in several years ago.

Pull request https://github.com/pytorch/pytorch/issues/142023 adds a static_assert that NCCL version should be 2.7 or above for PyTorch to work.
2.4 released about 2 years ago so it's relatively safe to assume that everyone has upgraded.

Assume that the world has all upgraded to later version of NCCL library.

Release note for PyTorch must specify that going forward, PyTorch will only work with NCCL version 2.7 and above.


Test Plan:
Unit tests.





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

Reviewed By: wconstab, fduwjj, kwen2501

Differential Revision: D66672062

Pulled By: c-p-i-o

kwen2501 commented:
Closing via https://github.com/pytorch/pytorch/pull/148900
